### ISSUE TITLE
[2.0.4] array binding bugs

### ISSUE THREAD

**Issue by ashpumpkin** [phalcon/cphalcon#10616]
Look at this code:
```PHP
<?php
use Phalcon\DI,
	Phalcon\Mvc\Model,
	Phalcon\Db\Adapter\Pdo\MySQL as Connection,
	Phalcon\Mvc\Model\Manager as ModelsManager;


$di = new DI();

$connection = new Connection(array(
	"host"     => "localhost",
	"username" => "root",
	"password" => "12345",
	"dbname"   => "phalcon_test",
));

$modelsManager = new ModelsManager();
$modelsManager->setDi($di);
$di['db'] = $connection;
$di['modelsManager'] = $modelsManager;
$di['modelsMetadata'] = new Model\MetaData\Memory();

class Robots extends Model
{
	public $id;
}

$di->get('db')->execute('DROP TABLE IF EXISTS robots;');
$di->get('db')->execute('CREATE TABLE `robots` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;');

$robot = new Robots();
$robot->id = 1;
$robot->save();
$robot = new Robots();
$robot->id = 2;
$robot->save();

$robots = Robots::find([
	'id IN ({id:array})',
	'bind' => [
		'id' => [1, 2]
	]
]);

echo 'found ' . $robots->count() . ' robots';
echo '<br>';
try {
	$robots = Robots::find([
		'id IN ({id:array})',
		'bind' => [
			'id' => [1 => 1, 2 => 2]
		]
	]);

	echo 'found ' . $robots->count() . ' robots';
} catch (\Exception $exception) {
	echo $exception->getMessage();
}
```

It outputs:

```
found 2 robots
SQLSTATE[HY093]: Invalid parameter number: parameter was not defined
```

I think this code should work well regardless of the array keys. Can somebody fix it?

**Comment by ashpumpkin** [comment]
I think i've just found another one.

```PHP
<?php
use Phalcon\DI,
	Phalcon\Mvc\Model,
	Phalcon\Db\Adapter\Pdo\MySQL as Connection,
	Phalcon\Mvc\Model\Manager as ModelsManager;


$di = new DI();

$connection = new Connection([
	"host"     => "localhost",
	"username" => "root",
	"password" => "12345",
	"dbname"   => "phalcon_test",
]);

$modelsManager = new ModelsManager();
$modelsManager->setDi($di);
$di['db'] = $connection;
$di['modelsManager'] = $modelsManager;
$di['modelsMetadata'] = new Model\MetaData\Memory();

class Robots extends Model
{
	public $id;
}

$di->get('db')->execute('DROP TABLE IF EXISTS robots;');
$di->get('db')->execute('CREATE TABLE `robots` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;');

$robot = new Robots();
$robot->id = 1;
$robot->save();
$robot = new Robots();
$robot->id = 2;
$robot->save();

$robots = Robots::find([
	'id IN ({id:array})',
	'bind' => [
		'id' => [1, 2]
	]
]);

echo 'found ' . $robots->count() . ' robots';
echo '<br>';

try {
	$robots = Robots::find([
		'id IN ({id:array})',
		'bind' => [
			'id' => [2]
		]
	]);

	echo 'found ' . $robots->count() . ' robots';
	echo '<br>';
} catch (\Exception $exception) {
	echo $exception->getMessage();
}
```

Outputs:
```
found 2 robots
SQLSTATE[HY093]: Invalid parameter number: parameter was not defined
```

:-(

**Comment by olivier-monaco** [comment]
It's a cache problem. The second call of find use the query of the first one.

**Comment by ashpumpkin** [comment]
@olivier-monaco the first bug is not related to cache.
```PHP
$robot = new Robots();
$robot->id = 1;
$robot->save();
$robot = new Robots();
$robot->id = 2;
$robot->save();

try {
	$robots = Robots::find([
		'id IN ({id:array})',
		'bind' => [
			'id' => [1 => 1, 2 => 2]
		]
	]);

	echo 'found ' . $robots->count() . ' robots';
} catch (\Exception $exception) {
	echo $exception->getMessage();
}
```

this code reproduces invalid param number too

**Comment by olivier-monaco** [comment]
Yes, it because you are using an associated array instead of an indexed array.

**Comment by ashpumpkin** [comment]
@olivier-monaco 
```PHP
$robot = new Robots();
$robot->id = 1;
$robot->save();
$robot = new Robots();
$robot->id = 2;
$robot->save();

$indexedArray = [1,2,3,4];
unset($indexedArray[0]);

try {
	$robots = Robots::find([
		'id IN ({id:array})',
		'bind' => [
			'id' => $indexedArray
		]
	]);

	echo 'found ' . $robots->count() . ' robots';
} catch (\Exception $exception) {
	echo $exception->getMessage();
}
```
=> SQLSTATE[HY093]: Invalid parameter number: parameter was not defined


Looks like phalcon iterates my array like this:
```PHP
for(i = 0; i < arrayLength; ++ i){/* do something with array[i] */}
```
And this causes the problem when some array keys are missing or not integer

**Comment by sszdh** [comment]
It seams Zephir has no `foreach` or any array specific looping functionslity! Just `for` & indexing!

**Comment by sszdh** [comment]
Oh, my bad... my last comment is wrong i think!

```php

let items = ["a": 1, "b": 2, "c": 3, "d": 4];

for key, value in items {
    echo key, " ", value, "\n";
}

// Or

for item in ["a", "b", "c", "d"] {
    echo item, "\n";
}

```

**Comment by aleksandrzen** [comment]
> It seams Zephir has no foreach or any array specific looping functionslity! Just for & indexing!

@sszdh, from [documentation](http://zephir-lang.com/control.html#loops):

> let items = ["a": 1, "b": 2, "c": 3, "d": 4];
```
for key, value in items {
    echo key, " ", value, "\n";
}
```

**Comment by ashpumpkin** [comment]
BTW, why is it necessary to use type placeholders for array binding? I tried "id IN (:roboid:)" and it didn't work :disappointed:

**Comment by olivier-monaco** [comment]
Cache problem is not so easy to resolve. It is located in Phalon\Mvc\Model\Query::parse. This method calls  phql_parse_phql to transform the PHQL statement to a PHQL AST and translate this AST to a intermediate SQL representation.

The phql_parse_phql function does no look for the bound parameter when parsing "id IN ({id:array})": it's only a semantic process. The resulting AST is the same for any bound array. It caches the resulting AST (using an hash of the PHQL query) and return the cached one is the same PHQL query is provided. This function associate an unique ID to each AST.

When transforming the PHQL AST to the intermediate SQL representation, the Query class use the bound parameter when processing an array placeholder (cf. line 606, inside _getExpression method). It keeps the size of the bound array in the intermediate SQL representation.

So one PHQL AST can give many intermediate SQL representation, depending on the bound array. Two problems exists:
1) Once the Query parsed the PHQL, it keeps the intermediate SQL representation. Executing the same query using two array of different length will through an error.
2) The parse method use the PHQL AST ID to cache the intermediate SQL representation. The previous problem is propagated to all Query created using the same PHQL query.

Usecase:
<pre>
<?php
$di = new Phalcon\DI();

$di['db'] = new Phalcon\Db\Adapter\Pdo\MySQL([
    "host"     => "localhost",
    "username" => "phalcon_test",
    "password" => "phalcon_test",
    "dbname"   => "phalcon_test",
]);

$di['modelsManager'] = new Phalcon\Mvc\Model\Manager();
$di['modelsManager']->setDi($di);
$di['modelsMetadata'] = new Phalcon\Mvc\Model\MetaData\Memory();

class Robot extends Phalcon\Mvc\Model
{
    public $id;
}

$q = new Phalcon\Mvc\Model\Query('select 1 in ({a:array}) from Robot', $di);

$q->setBindParams(['a' => [1]]);
$s = $q->getSql();
echo 'Bound array length: '.count($s['bind']['a']).PHP_EOL;
echo 'Expected "SELECT 1 IN (:a0) AS `_0` FROM `robot`"'.PHP_EOL;
echo 'Real     "'.$s['sql'].'"'.PHP_EOL;
echo PHP_EOL;

$q->setBindParams(['a' => [1, 2]]);
$s = $q->getSql();
echo 'Bound array length: '.count($s['bind']['a']).PHP_EOL;
echo 'Expected "SELECT 1 IN (:a0, :a1) AS `_0` FROM `robot`"'.PHP_EOL;
echo 'Real     "'.$s['sql'].'"'.PHP_EOL;
echo PHP_EOL;

$q = new Phalcon\Mvc\Model\Query('select 1 in ({a:array}) from Robot', $di);

$q->setBindParams(['a' => [1, 2, 3]]);
$s = $q->getSql();
echo 'Bound array length: '.count($s['bind']['a']).PHP_EOL;
echo 'Expected "SELECT 1 IN (:a0, :a1, :a2) AS `_0` FROM `robot`"'.PHP_EOL;
echo 'Real     "'.$s['sql'].'"'.PHP_EOL;
</pre>

Current output:
<pre>
Bound array length: 1
Expected "SELECT 1 IN (:a0) AS `_0` FROM `robot`"
Real     "SELECT 1 IN (:a0) AS `_0` FROM `robot`"

Bound array length: 2
Expected "SELECT 1 IN (:a0, :a1) AS `_0` FROM `robot`"
Real     "SELECT 1 IN (:a0) AS `_0` FROM `robot`"

Bound array length: 3
Expected "SELECT 1 IN (:a0, :a1, :a2) AS `_0` FROM `robot`"
Real     "SELECT 1 IN (:a0) AS `_0` FROM `robot`"
</pre>

I see 3 ways to resolve this issue:
1) The fast but bad way: use a preg_match, looking for "\{[^:}]:array\}" and disabling intermediate SQL representation caching is found. Bad because we disable cache and we use a text matching solution, "parsing" the PHQL query again.
2) The "hack" way: add a "no-cache" flag and set it to true when _getExpression use the bound parameters. Better than the first way but still bad because we disable the cache.
3) The best but "expensive" way: use the bound parameters as late as possible, in the _executeSelect method. In fact, change the Phalcon\Db\Dialect::select method (and insert, update, delete...) to accept the bound parameters as an optional parameter. This implies changing all the getSqlExpression methods (25+)...

@andresgutierrez What do you think?

**Comment by andresgutierrez** [comment]
@olivier-monaco Thanks for the analysis, opted for the third option

It's now implemented in the 2.0.x branch
